File: CPP/medium/brickWall.cpp
#include <vector>
#include <unordered_map>
#include <iostream>

int leastBricks(std::vector<std::vector<int>>& wall) {
    // memorize the every gap's position
    std::unordered_map<int, int> gap = {};
    int row = wall.size();
    int maxGap = 0;

    for(int i = 0; i < row; i++){
        // indicate the gap position
        int currentP = 0;
        // Since we cannot cut the through the edge of the wall,
        // loop through the entire row except the last element
        for(int brick = 0; brick < wall[i].size() - 1; brick++){
            currentP += wall[i][brick];
            // increment the gap position by 1 to indicate a gap
            // shows up at the current position
            if(gap.count(currentP)){
                gap[currentP]++;
            }
            else{
                gap.insert({currentP, 1});
            }
            if(gap[currentP] > maxGap){
                maxGap =  gap[currentP];
            }
        }
    }
    // the answer should be cutting through the position with the most gap
    return row - maxGap;
}

void printMatrix(std::vector<std::vector<int>> matrix){
    for(int row = 0; row < matrix.size(); row++){
        for(int col = 0; col < matrix[row].size(); col++){
            std::cout << (matrix[row][col] < 10?" ":"") << " " << matrix[row][col] ;
        }
        std::cout <<std::endl;
    }
    std::cout << std::endl;
}

// ============ Test ============
//   1  1
//   2
//   1  1

// The Minimum Number of Crossed Bricks After Drawing a Vertical Line = 1
// ============ Test ============
//   1  2  2  1
//   3  1  2
//   1  3  2
//   2  4
//   3  1  2
//   1  3  1  1

// The Minimum Number of Crossed Bricks After Drawing a Vertical Line = 2

int main(){
    std::vector<std::vector<std::vector<int>>> testSet = {
        {
            {1,1},
            {2},
            {1,1}
        },
        {
            {1,2,2,1},
            {3,1,2},
            {1,3,2},
            {2,4},
            {3,1,2},
            {1,3,1,1}
        }
    };
    for(auto test : testSet){
        std::cout << "============ Test ============" << std::endl;
        printMatrix(test);
        std::cout << "The Minimum Number of Crossed Bricks After Drawing a Vertical Line = "<< leastBricks(test) <<std::endl;
    }
    return 0;
}

File: Java/medium/brickWall.java
import java.util.*;

class Solution {
    public static int leastBricks(List<List<Integer>> wall) {
        HashMap<Integer, Integer> gap = new HashMap<>();

        int row = wall.size();
        int maxGap = 0;
        for(int i = 0; i < wall.size(); i++){
            List<Integer> currentRow = wall.get(i);
            int currentPosition = 0;
            for(int brick = 0; brick < currentRow.size() - 1; brick++){
                currentPosition += currentRow.get(brick);
                if(gap.get(currentPosition) != null){
                    gap.replace(currentPosition, gap.get(currentPosition) + 1);
                }
                else{
                    gap.put(currentPosition, 1);
                }
                if(maxGap < gap.get(currentPosition)){
                    maxGap = gap.get(currentPosition);
                }
            }
        }

        return row - maxGap;        
    }

    public static void printMatrix(List<List<Integer>> matrix){
        for(int row = 0; row < matrix.size(); row++){
            for(int col = 0; col < matrix.get(row).size(); col++){
                System.out.print((matrix.get(row).get(col) < 10?" ":"") + " " + matrix.get(row).get(col));
            }

            System.out.println();
        }
        System.out.println();
    }

//   ============ Test ============
//     1  1
//     2
//     1  1
  
//   The Minimum Number of Crossed Bricks After Drawing a Vertical Line = 1
  
//   ============ Test ============
//     1  2  2  1
//     3  1  2
//     1  3  2
//     2  3
//     3  1  2
//     1  3  1  1
  
//   The Minimum Number of Crossed Bricks After Drawing a Vertical Line = 2

    public static void main(String[] args){
        List<List<List<Integer>>> testSet = Arrays.asList(
            Arrays.asList(
                Arrays.asList(1,1),
                Arrays.asList(2),
                Arrays.asList(1,1)
            ),
            Arrays.asList(
                Arrays.asList(1,2,2,1),
                Arrays.asList(3,1,2),
                Arrays.asList(1,3,2),
                Arrays.asList(2,3),
                Arrays.asList(3,1,2),
                Arrays.asList(1,3,1,1)
            )
        );
        
        for(List<List<Integer>> test : testSet){
            System.out.println("============ Test ============");
            printMatrix(test);
            System.out.println("The Minimum Number of Crossed Bricks After Drawing a Vertical Line = " + leastBricks(test) + '\n');
        }
    
    }
}